core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -99,11 +103,31 @@
   private int requiredLength = HEADER_LENGTH;
   private boolean compressedFlag;
   private boolean endOfStream;
-  private CompositeReadableBuffer nextFrame;
   private CompositeReadableBuffer unprocessed = new CompositeReadableBuffer();
-  private long pendingDeliveries;
+  private ArrayDeque<Frame> appQueue;
+
+  // TODO: Need to fix false sharing issues
+  private final AtomicReference<ArrayDeque<Frame>> queueRef =
+      new AtomicReference<ArrayDeque<Frame>>();
+
   private boolean deliveryStalled = true;
-  private boolean inDelivery = false;
+
+  // The number of undelivered messages requested by the application.
+  private long requestedMessageCount;
+  private boolean messagesAvailable;
+  // Used to track whether the application is making re-entrant calls to request.
+  private boolean appInRequest;
+  // The number of messages requested by the application observed by the transport thread
+  // used to optimize return of flow-control.
+  private AtomicLong transportCount = new AtomicLong(0);
+
+  // Number of bytes received from the transport
+  private long bytesReceived;
+  // Number of bytes received from the transport that have been consumed.
+  private long bytesConsumed;
+  // Number of flow-control bytes that have been returned to the transport either immediately
+  // or via the queue of events processed by the application.
+  private long bytesReturned;
```

Comment:
It seems that `bytesConsumed` and `bytesReturned` continually increase as more data is read on the stream ... never decreasing.  I'm concerned about overflow for large, long-running streams.